src/fft.rg

```diff
@@ -13,74 +13,135 @@
+    elseif result == 4 then
+      status = "CUFFT_INVALID_VALUE"
+    elseif result == 5 then
+      status = "CUFFT_INTERNAL_ERROR"
+    elseif result == 6 then
+      status = "CUFFT_EXEC_FAILED"
+    elseif result == 7 then
+      status = "CUFFT_SETUP_FAILED"
+    elseif result == 8 then
+      status = "CUFFT_INVALID_SIZE"
+    elseif result == 9 then
+      status = "CUFFT_UNALIGNED_DATA"
+    elseif result == 10 then
+      status = "CUFFT_INCOMPLETE_PARAMETER_LIST"
+    elseif result == 11 then
+      status = "CUFFT_INVALID_DEVICE"
+    elseif result == 12 then
+      status = "CUFFT_PARSE_ERROR"
+    elseif result == 13 then
+      status = "CUFFT_NO_WORKSPACE"
+    elseif result == 14 then
+      status = "CUFFT_NOT_IMPLEMENTED"
+    elseif result == 15 then
+      status = "CUFFT_LICENSE_ERROR"
+    elseif result == 16 then
+      status = "CUFFT_NOT_SUPPORTED"
+    end
+    regentlib.assert(result == cufft_c.CUFFT_SUCCESS, status)
+  end
 end
 
--- Hack: get defines from fftw3.h
+-- Define constants: get defines from fftw3.h
 fftw_c.FFTW_FORWARD = -1
 fftw_c.FFTW_BACKWARD = 1
-
 fftw_c.FFTW_MEASURE = 0
-fftw_c.FFTW_DESTROY_INPUT = (2 ^ 0)
-fftw_c.FFTW_UNALIGNED = (2 ^ 1)
-fftw_c.FFTW_CONSERVE_MEMORY = (2 ^ 2)
-fftw_c.FFTW_EXHAUSTIVE = (2 ^ 3) -- NO_EXHAUSTIVE is default
-fftw_c.FFTW_PRESERVE_INPUT = (2 ^ 4) -- cancels FFTW_DESTROY_INPUT
-fftw_c.FFTW_PATIENT = (2 ^ 5) -- IMPATIENT is default
 fftw_c.FFTW_ESTIMATE = (2 ^ 6)
-fftw_c.FFTW_WISDOM_ONLY = (2 ^ 21)
 
 local fft = {}
 
-function fft.generate_fft_interface(itype, dtype)
+--- Create the FFT interface.
+-- @param itype Index type of transform (int2d/int2d/int3d/int4d).
+-- @param dtype_in Input data type of transform (float/double/complex32/complex64).
+-- @param dtype_out Output data type of transform (complex32/complex64).
+function fft.generate_fft_interface(itype, dtype_in, dtype_out)
   assert(regentlib.is_index_type(itype), "requires an index type as the first argument")
   local dim = itype.dim
-  assert(dim >= 1 and dim <= 3, "currently only 1 <= dim <= 3 is supported")
-  assert(dtype == complex64, "currently only complex64 is supported")
+  local dtype_out_size = terralib.sizeof(dtype_out)
+  assert(dim >= 1 and dim <= 4, "currently only 1 <= dim <= 4 is supported")
+  assert(dtype_in == float or dtype_in == double or dtype_in == complex32 or dtype_in == complex64, "input type must be float/double/complex32/complex64")
+  assert(dtype_out == complex32 or dtype_out == complex64, "output type must be complex32/complex64")
+
+   -- Single-precision transforms
+  local float_to_complex32_transform = (dtype_in == float and dtype_out == complex32)
+  local complex32_to_complex32_transform = (dtype_in == complex32 and dtype_out == complex32)
+
+   -- Double-precision transforms
+  local double_to_complex64_transform = (dtype_in == double and dtype_out == complex64)
+  local complex64_to_complex64_transform = (dtype_in == complex64 and dtype_out == complex64)
```

Comment:
This is not the way I would want to do this. It appears that you are making variables here for the different types, and then doing *dynamic* branches in the code. That means that every interface will link to every FFT function as we do not differentiate them at runtime. It also means we always end up linking both double and single precision FFTW libraries even if we don't use both.

To me it would seem simpler to do:

```lua
local cufft_transform_type
if dtype_in == float and dtype_out == complex32 then
  cufft_transform_type = cufft_c.CUFFT_R2C
elseif ... then
  ...
else
  assert(false, "Unexpected type combination " .. tostring(dtype_in) .. " and " .. tostring(dtype_out))
end
```

And then later where you have the if statement you just:

```lua
cufft_assert(cufft_c.cufftPlanMany(&p.cufft_p, dim, &n[0], [&int](0), 0, 0, [&int](0), 0, 0, cufft_transform_type, 1))
```

No `if` statement because you branched on the FFT type earlier.

You can do the same for FFTW even though it uses different functions:

```lua
local fftw_create_plan, fftw_complex
if dtype_in == float and dtype_out == complex32 then
  fftw_create_plan = fftw_c.fftwf_plan_dft_r2c
  fftw_complex = fftw_c.fftwf_complex
elseif ... then
  ...
else
  assert(false, "Unexpected type combination " .. tostring(dtype_in) .. " and " .. tostring(dtype_out))
end
```

And then when you use it:

```lua
fftw_create_plan(dim, &n[0], [&float](input_base), [&fftw_complex](output_base), fftw_c.FFTW_ESTIMATE)
```

You can meta-program the type of the field `p` so that you don't need multiple plan fields either.

Reply:
Have implemented this - see commit https://github.com/arjunkunna/regent-fft/commit/cf2dd5c77c68227b8083c38d5665eb067cf78c55